usr/lib/common/asn1.c

```diff
@@ -3473,3 +3574,806 @@ CK_RV ber_decode_DHPublicKey(CK_BYTE *data,
+    }
+
+    /* Build t1 attribute */
+    rc = build_attribute(CKA_IBM_DILITHIUM_T1, t1, t1_len, &t1_attr_temp);
+    if (rc != CKR_OK) {
+        TRACE_DEVEL("build_attribute failed\n");
+        goto cleanup;
+    }
+
+    *rho_attr = rho_attr_temp;
+    *t1_attr = t1_attr_temp;
+
+    return CKR_OK;
+
+cleanup:
+    if (rho_attr_temp)
+        free(rho_attr_temp);
+    if (t1_attr_temp)
+        free(t1_attr_temp);
+
+    return rc;
+}
+
+/**
+ * An IBM Dilithium private key is given by:
+ *
+ *     DilithiumPrivateKey ::= SEQUENCE {
+ *       version INTEGER,     -- v0, reserved 0
+ *       rho BIT STRING,      -- nonce
+ *       key BIT STRING,      -- key/seed/D
+ *       tr  BIT STRING,      -- PRF bytes ('CRH' in spec)
+ *       s1  BIT STRING,      -- vector(L)
+ *       s2  BIT STRING,      -- vector(K)
+ *       t0  BIT STRING       -- low bits(vector L)
+ *       t1 [0] IMPLICIT OPTIONAL {
+ *         t1  BIT STRING     -- high bits(vector L)  -- see also public key
+ *       }
+ *     }
+ */
+CK_RV ber_encode_IBM_DilithiumPrivateKey(CK_BBOOL length_only,
+                               CK_BYTE **data,
+                               CK_ULONG *data_len,
+                               CK_ATTRIBUTE *rho,
+                               CK_ATTRIBUTE *seed,
+                               CK_ATTRIBUTE *tr,
+                               CK_ATTRIBUTE *s1,
+                               CK_ATTRIBUTE *s2,
+                               CK_ATTRIBUTE *t0,
+                               CK_ATTRIBUTE *t1,
+                               CK_ATTRIBUTE *opaque)
+{
+    CK_BYTE *buf = NULL;
+    CK_BYTE *buf2 = NULL;
+    CK_ULONG len, offset;
+    CK_BYTE version[] = { 0 };
+    CK_RV rc;
+    BerElement *ber;
+    BerValue *val;
+
+    /* Calculate storage for sequence */
+    offset = 0;
+    rc = 0;
+
+    rc |= ber_encode_INTEGER(TRUE, NULL, &len, NULL, sizeof(version));
+    offset += len;
+    if (opaque != NULL) {
+        rc |= ber_encode_BIT_STRING(TRUE, NULL, &len, NULL, opaque->ulValueLen, 0);
+        offset += len;
+    } else {
+        rc |= ber_encode_BIT_STRING(TRUE, NULL, &len, NULL, rho->ulValueLen, 0);
+        offset += len;
+        rc |= ber_encode_BIT_STRING(TRUE, NULL, &len, NULL, seed->ulValueLen, 0);
+        offset += len;
+        rc |= ber_encode_BIT_STRING(TRUE, NULL, &len, NULL, tr->ulValueLen, 0);
+        offset += len;
+        rc |= ber_encode_BIT_STRING(TRUE, NULL, &len, NULL, s1->ulValueLen, 0);
+        offset += len;
+        rc |= ber_encode_BIT_STRING(TRUE, NULL, &len, NULL, s2->ulValueLen, 0);
+        offset += len;
+        rc |= ber_encode_BIT_STRING(TRUE, NULL, &len, NULL, t0->ulValueLen, 0);
+        offset += len;
+        if (t1) {
+            ber = ber_alloc_t(LBER_USE_DER);
+            rc = ber_put_bitstring(ber, t1->pValue, t1->ulValueLen * 8, 0x03);
+            rc = ber_flatten(ber, &val);
+
+            ber_encode_CHOICE(TRUE, 0, NULL, &len, (CK_BYTE *)val->bv_val,
```

Comment:
error checking ?

Reply:
Yes, to get a rc = 0 only if all operations succeeded, we must use rc |= as above.

Reply:
Interesting: ber_put_bitstring returns the number of put bits, so a check for rc = 0 fails. I guess this is the reason why it was always coded without rc checking. I will check if we can replace ber_put_bitstring by ber_encode_BIT_STRING, which would also remove this dependency to OpenLDAP.

Reply:
OK, I can replace ber_put_bitstring / ber_flatten by ber_encode_BIT_STRING. So I will do this now for Dilithium, but let's not forget to fix the other places also to get rid of OpenLdap.

Reply:
ber = ber_alloc_t(LBER_USE_DER); above also needs  a check for NULL i guess.

Reply:
Instead of looking to replace -lber i would rather see if we can replace our own ber stuff by lber or some other ber encoding library.